src/hashtable-shim/hash-table.lisp

```diff
@@ -0,0 +1,321 @@
+          :keys (make-array capacity :element-type 't)
+          :metadata (make-array capacity :element-type '(signed-byte 8) :initial-element +slot-empty+)
+          :size 0
+          :hash-function hash-function
+          :eq-function eq-function)))
+
+(declaim (inline hash-matches))
+(defun hash-matches% (metadata keys key short-hash index eq-function)
+  "Returns t if SHORT-HASH and HASH are matches in the arrays METADATA and KEYS at INDEX."
+  (declare (type (simple-array (signed-byte 8) (*)) metadata)
+           (type (simple-array t (*)) keys)
+           (type t key)
+           (type (signed-byte 8) short-hash)
+           (type fixnum index)
+           (type (function (t t) boolean) eq-function)
+           (values boolean)
+           #.*hash-table-optimize*)
+  (and (= short-hash (aref metadata index))
+       (funcall eq-function key (aref keys index))))
+
+(defmacro across-hash-table ((index-var start capacity-var) &body body)
+  `(progn
+     (loop :for ,index-var fixnum :from ,start :below ,capacity-var
+           :do (progn ,@body))
+
+     (loop :for ,index-var fixnum :from 0 :below ,start
+           :do (progn ,@body))))
+
+(declaim (inline hash-table-insert%))
+(defun hash-table-insert% (metadata keys values capacity key element hash-function eq-function)
+  "Returns t if a new slot was used."
+  (declare (type (simple-array (signed-byte 8) (*)) metadata)
+           (type (simple-array t (*)) keys)
+           (type (simple-array t (*)) values)
+           (type fixnum capacity)
+           (type t key)
+           (type t element)
+           (type (function (t) hash) hash-function)
+           (type (function (t t) boolean) eq-function)
+           (values boolean &optional)
+           #.*hash-table-optimize*)
+  (let* ((hash (funcall hash-function key))
+
+         (short-hash (ldb (byte 7 0) hash))
+
+         (idx (rem hash capacity)))
+
+    (across-hash-table (i idx capacity)
+      ;; When encountering an empty slot, take it updating the slots
+      ;; metadata and key in the process. Also increment the tables
+      ;; size
+      (when (= +slot-empty+ (aref metadata i))
+        (setf (aref metadata i) short-hash)
+        (setf (aref keys i) key)
+        (setf (aref values i) element)
+        (return-from hash-table-insert% t))
+
+      ;; If the slot has the same hash then only update the element
+
+      (when (hash-matches% metadata keys key short-hash i eq-function)
+        (setf (aref values i) element)
+        (return-from hash-table-insert% nil)))
+
+    (util:coalton-bug "hash table full")))
+
+(defun hash-table-insert (table key element)
+  "Insert or update the values stored at KEY in TABLE."
+  (declare (type hash-table table)
+           (type t key)
+           (type t element)
+           (values null)
+           #.*hash-table-optimize*)
+
+  (when (>= (the fixnum (* (hash-table-size table) 10))
+            (the fixnum (* (the fixnum (hash-table-capacity table)) 7))) 
```

Comment:
Are you comfortable with these magic numbers?

Reply:
Added a comment